### PR TITLE
feat(stages): recover senders for both DB and static file transactions

### PR DESCRIPTION
**Pull request by shekhirin**
With this PR, when recovering senders in the `SenderRecovery` pipeline stage, we will look into both database table `Transactions` and static files for `Transactions` segment. It becomes relevant when https://github.com/paradigmxyz/reth/pull/5733 is merged.

One thing to note with the current implementation is that we no longer retrieve `RawKey` and `RawValue` from the database when querying transactions for sender recovery, but instead decode them before passing the transactions to threads for recovery. The bench showed there's no significant difference in performance between the two approaches. 

Holesky benchmark (`main` vs this PR):
```console
ubuntu@reth4:~/reth-snaps-alexey$ hyperfine './reth_main stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders'
Benchmark 1: ./reth_main stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders
  Time (mean ± σ):     48.451 s ±  0.335 s    [User: 1696.042 s, System: 6.879 s]
  Range (min … max):   47.847 s … 48.946 s    10 runs

ubuntu@reth4:~/reth-snaps-alexey$ hyperfine './reth_6089_raw stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders'
Benchmark 1: ./reth_6089_raw stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders
  Time (mean ± σ):     52.559 s ±  0.146 s    [User: 1762.202 s, System: 8.185 s]
  Range (min … max):   52.248 s … 52.708 s    10 runs
```

---

In addition to that, we need to keep in mind that now transactions, headers, and receipts can be in static files, so when calculating the stage checkpoint, we need to look in both database and static files. I introduced a trait `StatsReader` for that, and replaced all calls to `tx.entries::<T>()` with it.

### CODE REVIEW

**Review by onbjerg — APPROVED** [review]

**Comment by shekhirin** [comment]
Okay, it performs worse... I need to return the raw inserts functionality

```console
ubuntu@reth4:~/reth-snaps-alexey$ hyperfine './reth_main stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders'
Benchmark 1: ./reth_main stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders
  Time (mean ± σ):     48.451 s ±  0.335 s    [User: 1696.042 s, System: 6.879 s]
  Range (min … max):   47.847 s … 48.946 s    10 runs

ubuntu@reth4:~/reth-snaps-alexey$ hyperfine './reth_6089 stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders'
Benchmark 1: ./reth_6089 stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders
  Time (mean ± σ):     64.176 s ±  0.451 s    [User: 1700.714 s, System: 15.671 s]
  Range (min … max):   63.374 s … 64.920 s    10 runs
```

**Comment by shekhirin** [comment]
Raw query looks good

```console
ubuntu@reth4:~/reth-snaps-alexey$ hyperfine './reth_main stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders'
Benchmark 1: ./reth_main stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders
  Time (mean ± σ):     48.451 s ±  0.335 s    [User: 1696.042 s, System: 6.879 s]
  Range (min … max):   47.847 s … 48.946 s    10 runs

ubuntu@reth4:~/reth-snaps-alexey$ hyperfine './reth_6089_raw stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders'
Benchmark 1: ./reth_6089_raw stage --chain holesky run --datadir ~/.local/share/reth/holesky --from 0 --to 808004 senders
  Time (mean ± σ):     52.559 s ±  0.146 s    [User: 1762.202 s, System: 8.185 s]
  Range (min … max):   52.248 s … 52.708 s    10 runs
```

Anyway, will remove the querying for snapshots at all for now, because we don't write to them yet.

File: crates/stages/benches/criterion.rs [hunk 1]
```diff
@@ -95,44 +96,58 @@ fn merkle(c: &mut Criterion) {
...
+    let db = setup::txs_testdata(DEFAULT_NUM_BLOCKS);
+
     let stage = MerkleStage::Both { clean_threshold: u64::MAX };
     measure_stage(
         &mut group,
+        &db,
         setup::unwind_hashes,
         stage,
-        1..DEFAULT_NUM_BLOCKS,
+        1..=DEFAULT_NUM_BLOCKS,
```
Comment (onbjerg) [inline crates/stages/benches/criterion.rs:107]: was this a bug before?

**Review by onbjerg — APPROVED** [review]
lgtm

File: crates/stages/benches/criterion.rs (hunk 1 above)
Comment (onbjerg) [reply crates/stages/benches/criterion.rs:107]: or well "bug", i guess this is in a benchmark, so doesn't rly matter

Comment (shekhirin) [reply crates/stages/benches/criterion.rs:107]: the behavior didn't change: even though we were passing an exclusive range, we constructed `StageRange` struct with `range.start` and `range.end`

**Review by onbjerg — APPROVED** [review]
lgtm

**Review by emhane — COMMENTED** [review]
I like that you don't constraint generics before it's necessary, it improves readability of the code. it also makes the code more flexible, and therefore easier to test since generic types without constraints can easily be mocked.

in some places the code can panic by calling `expect`. in these situations, are you sure you want the node to stop execution? I'd suggest using `debug_assert` and dumping as much state as could be helpful to debug to log in the message. instead in run mode, consider if it's better to conditionally continue execution by `if let Some(..) ..` and `if let Ok(..) ..`. in some cases it may indeed be best for the node to stop.

File: crates/stages/src/stages/sender_recovery.rs [hunk 2]
```diff
@@ -188,14 +186,11 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
...
 
 fn recover_sender(
-    entry: Result<(RawKey<TxNumber>, RawValue<TransactionSignedNoHash>), DatabaseError>,
+    (tx_id, tx): (TxNumber, RawValue<TransactionSignedNoHash>),
     rlp_buf: &mut Vec<u8>,
 ) -> Result<(u64, Address), Box<SenderRecoveryStageError>> {
-    let (tx_id, transaction) =
-        entry.map_err(|e| Box::new(SenderRecoveryStageError::StageError(e.into())))?;
-    let tx_id = tx_id.key().expect("key to be formated");
+    let tx = tx.value().expect("value to be formated");
```
Comment (emhane) [inline crates/stages/src/stages/sender_recovery.rs:192]: for example here, could it be possible that one tx value is not formatted but other tx values are? in that case, it's better not to panic

Comment (emhane) [reply crates/stages/src/stages/sender_recovery.rs:192]: alright, in case of a bug, would be nice to get more context, same advice you gave me last week

Comment (emhane) [reply crates/stages/src/stages/sender_recovery.rs:192]: btw, here, and in tracing, it's nice if the variable names match the log message keys exactly. makes it easier to debug. so, 
```rust
let my_var = 1;
debug!(target: "..",
    my_var=my_var,
    ""
);
```

**Comment by shekhirin** [comment]
ETA bug is unrelated to these changes, the progress is calculated correctly

**Pull request merged**